MPI Communication

sslattery opened the pull request:
Adds support for MPI and basic communication plans and operations:

1. `CommunicationPlan` - base class for all communication plans. Contains the base implementation which determines the number of imports and exports as well as neighbor ranks based on export data.

2. `Distributor` - provides a communication plan for `migrate` to move data from one uniquely-owned distribution to another uniquely-owned distribution. This derives from `CommunicationPlan`.

3. `Halo` - provides a communication plan for `gather` and `scatter` to manipulate ghosted data. This also derives from `CommunicationPlan`

Each has a decent set of unit tests that caught some bugs. I have also tested this with a GPU-aware OpenMPI implementation although we are currently only using CUDA UVM (we will be adding regular CUDA memory soon).

sslattery commented:
@dalg24 @junghans @sjplimp @rfbird This one is finally ready for review. I am not opposed to splitting this PR into 3 PRs - one for each of the 3 new classes. Please take a look when you have a chance. Feedback on the APIs would be useful.

junghans commented on .travis.yml:
```diff
@@ -47,6 +47,7 @@ script:
     cmake -DCMAKE_PREFIX_PATH=$HOME/kokkos
           -DCabana_ENABLE_Serial=OFF ${CMAKE_OPTS[@]}
           -DCabana_ENABLE_MPI=ON
+          -DMPIEXEC_EXECUTABLE=/usr/bin/mpirun
```
Isn't that automatic?

sslattery replied:
Suprisingly it didn't work. On my other machines the CMake MPI module automatically set this variable but it was not being defined in the travis environment. Adding it fixed the problem.

junghans replied:
Hmm, the CMake on travis is very new, so maybe the variable got renamed!

sslattery replied:
That would seem odd as I got my instructions from the latest docs: https://cmake.org/cmake/help/latest/module/FindMPI.html

junghans replied:
I will have a look!

junghans replied:
Got it, Travis run CMake-3.9 and it is called `MPIEXEC` (not `MPIEXEC_EXECUTABLE`) in that version.

junghans replied:
`MPIEXEC` is listed under [Backward-Compatibility](https://cmake.org/cmake/help/v3.11/module/FindMPI.html#backward-compatibility), so we should use that one.

dalg24 commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+            {
+                _neighbors.push_back( status[i].MPI_SOURCE );
+                _num_import.push_back( import_sizes[i] );
+            }
```
How about something like
```C++
auto const source = status[i].MPI_SOURCE;
auto result = std::find(_neighbors.begin(), _neighbors.end(), source);
if (result != std::end(_neighbors))
{
    _num_import[i+self_offset] = import_sizes[i];
}
else
{
    _neighbors.push_back( source );
    _num_import.push_back( import_sizes[i] );
}
```

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+                                         _neighbors[n] );
+
+        // Post receives for the number of imports we will get.
+        std::vector<MPI_Request> requests(0);
```
👍

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+
+        // Wait on receives.
+        std::vector<MPI_Status> status( requests.size() );
+        MPI_Waitall( requests.size(), requests.data(), status.data() );
```
👍

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+        for ( int n = 0; n < num_n; ++n )
+            if ( my_rank != _neighbors[n] )
+            {
+                requests.resize( requests.size() + 1 );
```
👍

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+        // Extract the export ranks and number of exports and then flag the
+        // send ranks.
+        _neighbors.resize( 0 );
+        _num_export.resize( 0 );
```
👍

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+        std::vector<std::size_t> counts( num_n, 0 );
+        Kokkos::View<std::size_t*,Kokkos::HostSpace>
+            host_steering( "host_steering", _total_num_export );
+        for ( std::size_t i = 0; i < element_export_ranks.size(); ++i )
```
I made this parallel

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+        const bool use_iota,
+        const std::vector<int>& element_export_ranks,
+        const std::vector<std::size_t>& element_export_ids =
+        std::vector<std::size_t>() )
```
I refactored these functions to use overloading with different signatures

sslattery commented:
@dalg24 I refactored the inheritance structure here. The base class now has functions that are essentially protected but they must be made public to use class data with CUDA. I changed the create functions in the derived classes as well. This makes it clear that an entirely new communication plan is being created.

sslattery commented on core/src/Cabana_Distributor.hpp:
```diff
@@ -0,0 +1,503 @@
+    const Distributor_t& distributor,
+    const AoSoA_t& src,
+    AoSoA_t& dst,
+    int mpi_tag,
```
There are default arguments in the functions the user will call - the implementation detail assumes it will be passed a tag by these functions.

sslattery commented on core/src/Cabana_Distributor.hpp:
```diff
@@ -0,0 +1,503 @@
+        throw std::runtime_error("Destination is the wrong size for migration!");
+
+    // Get the number of components in the slice. The source and destination
+    // should be the same.
```
This is explicitly enforced by them having the same type `Slice_t` (i.e. the number of components is built into the type definition).

dalg24 replied:
Then say "is the same" instead of "should".

sslattery commented on core/src/Cabana_Halo.hpp:
```diff
@@ -0,0 +1,557 @@
+        if ( element_export_ids.size() != element_export_ranks.size() )
+            throw std::runtime_error("Export ids and ranks different sizes!");
+
+        _num_local = num_local;
```
it is used in `numLocal()` which in turn is used in `scatter` and `gather`. The data is laid out such that the local data comes first in the AoSoA/Slice and the ghosts come second.

sslattery commented on core/unit_test/tstCommunicationPlan.hpp:
```diff
@@ -0,0 +1,461 @@
+    {
+        this->createExportSteering(
+            element_export_ranks, element_export_ids );
+    }
```
yeah I had made this change so that they could be called in the unit test assuming they would be `protected`. Once I got on a GPU machine to test I realized they would need to be public and therefore hadn't changed this back to directly calling the object.

sslattery commented:
@junghans What do you think is the best way here to enable `CommunicationPlan`, `Distributor`, `Halo`, and their tests/examples only when MPI is enabled?

sslattery commented on core/src/Cabana_DeepCopy.hpp:
```diff
@@ -70,7 +70,6 @@ inline void deep_copy(
     // Return if both pointers are null.
     if ( dst_data == nullptr && src_data == nullptr )
     {
-        Kokkos::fence();
```
I removed this from the PR

sslattery commented on core/src/Cabana_Slice.hpp:
```diff
@@ -499,8 +499,6 @@ class Slice
 struct is_slice : public std::false_type {};
 
-// True only if the type is a member slice *AND* the member slice is templated
-// on an AoSoA type.
```
I removed this from the PR

junghans commented on core/unit_test/CMakeLists.txt:
```diff
@@ -44,7 +44,7 @@ endforeach()
 ## MPI tests.
 ##--------------------------------------------------------------------------##
-if(Cabana_ENABLE_MPI)
+if(${Cabana_ENABLE_MPI})
```
That is not needed!

junghans replied:
You can just use `if(Cabana_ENABLE_MPI)`.

sslattery replied:
I fixed this

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+                if ( element_export_ranks[i] == _neighbors[n] )
+                {
+                    host_steering( offsets[n] + counts[n] ) =
+                        (use_iota) ? i : element_export_ids[i];
```
should we do this now or leave for future optimizations?

dalg24 commented on core/src/CMakeLists.txt:
```diff
@@ -2,21 +2,57 @@ configure_file(CabanaCore_config.hpp.cmakein CabanaCore_config.hpp)
+  Cabana_Types.hpp
+  Cabana_VerletList.hpp
+  Cabana_Version.hpp
+  )
```
Say I create a new header and forget to list it here.  Will it be caught when building and running the tests?

sslattery replied:
It may although that is tough to tell because it is header only so it may just pull it in. It will likely show up more when external libraries link in and the missing header is not installed.

junghans replied:
No, it won’t, hence globbing and subfolders might be preferred!

sslattery replied:
Lets push this as is with regards to the build system them and move this conversation to #55 - once we come to a conclusion there we can implement in a new PR

dalg24 commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,602 @@
+      \return The number of export elements.
+    */
+    std::size_t numExportElement() const
+    { return _num_export_element; }
```
I understand we need something like this to validate the user input when executing the communication plan but I don't like the name of this member function because I don't find it very descriptive.

sslattery replied:
I have changed the name as we discussed

sslattery commented:
Things to finish:

- [ ] Fix name of `exportSteering()` -> `getExportSteering()`

- [ ] Fix name of `numExportElement()` -> `exportSize()`

- [ ] Add MPI_Barrier to all communication functions to avoid potential race conditions for same MPI tag calls and the add `async` versions in the future

sslattery commented on core/src/Cabana_CommunicationPlan.hpp:
```diff
@@ -0,0 +1,460 @@
+      elements going to neighbor with local id 1, etc.). Only indices of ghost
+      elements are in the list of exports.
+    */
+    Kokkos::View<std::size_t*,kokkos_memory_space> exportSteering() const
```
I have changed the name as we discussed
